File: src/main/java/ch/epfl/dias/calcite/adapter/pelago/metadata/PelagoRelMetadataQuery.java
package ch.epfl.dias.calcite.adapter.pelago.metadata;

import org.apache.calcite.plan.RelOptCost;
import org.apache.calcite.rel.RelDistribution;
import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rel.metadata.JaninoRelMetadataProvider;
import org.apache.calcite.rel.metadata.RelMetadataQuery;

import com.google.common.base.Preconditions;

import ch.epfl.dias.calcite.adapter.pelago.RelComputeDevice;
import ch.epfl.dias.calcite.adapter.pelago.RelDeviceType;
import ch.epfl.dias.calcite.adapter.pelago.RelHetDistribution;
import ch.epfl.dias.calcite.adapter.pelago.RelHomDistribution;
import ch.epfl.dias.calcite.adapter.pelago.RelPacking;

public class PelagoRelMetadataQuery extends RelMetadataQuery {
  private DeviceType.Handler      deviceTypeHandler;
  private Packing.Handler         packingHandler;
//  private SelfCost.Handler        selfCostHandler;
  private HetDistribution.Handler hetDistrHandler;
  private HomDistribution.Handler homDistrHandler;
  private ComputeDevice.Handler   computeTypeHandler;

  protected PelagoRelMetadataQuery(JaninoRelMetadataProvider metadataProvider,
                                   RelMetadataQuery prototype) {
    super(metadataProvider, prototype);
    deviceTypeHandler  = initialHandler(DeviceType.Handler.class);
    packingHandler     = initialHandler(Packing   .Handler.class);
//    selfCostHandler    = initialHandler(SelfCost  .Handler.class);
    hetDistrHandler    = initialHandler(HetDistribution.Handler.class);
    homDistrHandler    = initialHandler(HomDistribution.Handler.class);
    computeTypeHandler = initialHandler(ComputeDevice  .Handler.class);
  }

  public static PelagoRelMetadataQuery instance() {
    return new PelagoRelMetadataQuery(THREAD_PROVIDERS.get(), EMPTY);
  }
  /**
   * Returns the
   * {@link DeviceType#deviceType()}
   * statistic.
   *
   * @param rel         the relational expression
   */
  public RelDeviceType deviceType(RelNode rel) {
    for (;;) {
      try {
        return deviceTypeHandler.deviceType(rel, this);
      } catch (JaninoRelMetadataProvider.NoHandler e) {
        deviceTypeHandler = revise(e.relClass, DeviceType.DEF);
      }
    }
  }

  /**
   * Returns the
   * {@link DeviceType#deviceType()}
   * statistic.
   *
   * @param rel         the relational expression
   */
  public RelHetDistribution hetDistribution(RelNode rel) {
    for (;;) {
      try {
        return hetDistrHandler.hetDistribution(rel, this);
      } catch (JaninoRelMetadataProvider.NoHandler e) {
        hetDistrHandler = revise(e.relClass, HetDistribution.DEF);
      }
    }
  }

  public RelHomDistribution homDistribution(RelNode rel) {
    for (;;) {
      try {
        return homDistrHandler.homDistribution(rel, this);
      } catch (JaninoRelMetadataProvider.NoHandler e) {
        homDistrHandler = revise(e.relClass, HomDistribution.DEF);
      }
    }
  }

  public RelDistribution distribution(RelNode rel) {
    assert(false);
    return null;
  }

  /**
   * Returns the
   * {@link DeviceType#deviceType()}
   * statistic.
   *
   * @param rel         the relational expression
   */
  public RelComputeDevice computeType(RelNode rel) {
    for (;;) {
      try {
        return computeTypeHandler.computeType(rel, this);
      } catch (JaninoRelMetadataProvider.NoHandler e) {
        computeTypeHandler = revise(e.relClass, ComputeDevice.DEF);
      }
    }
  }

  /**
   * Returns the
   * {@link Packing#packing()}
   * statistic.
   *
   * @param rel         the relational expression
   */
  public RelPacking packing(RelNode rel) {
    for (;;) {
      try {
        return packingHandler.packing(rel, this);
      } catch (JaninoRelMetadataProvider.NoHandler e) {
        packingHandler = revise(e.relClass, Packing.DEF);
      }
    }
  }

//  /**
//   * Returns the
//   * {@link SelfCost#selfCost()}
//   * .
//   *
//   * @param rel         the relational expression
//   */
//  public RelOptCost selfCost(RelNode rel) {
//    for (;;) {
//      try {
//        return selfCostHandler.(rel, this);
//      } catch (JaninoRelMetadataProvider.NoHandler e) {
//        selfCostHandler = revise(e.relClass, SelfCost.DEF);
//      }
//    }
//  }
}

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/metadata/PelagoRelMetadataQuery.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */

package ch.epfl.dias.calcite.adapter.pelago.metadata

import ch.epfl.dias.calcite.adapter.pelago.traits._
import org.apache.calcite.rel.{RelDistribution, RelNode}
import org.apache.calcite.rel.metadata.RelMetadataQueryBase.initialHandler
import org.apache.calcite.rel.metadata.{
  ChainedRelMetadataProvider,
  JaninoRelMetadataProvider,
  RelMetadataQuery,
  RelMetadataQueryBase
}

import java.util

object PelagoRelMetadataQuery {
  def instance: PelagoRelMetadataQuery = {
    RelMetadataQueryBase.THREAD_PROVIDERS.set(
      JaninoRelMetadataProvider.of(
        ChainedRelMetadataProvider.of(
          util.List.of(
            RelMetadataQueryBase.THREAD_PROVIDERS.get,
            PelagoRelMetadataProvider.INSTANCE
          )
        )
      )
    )
    new PelagoRelMetadataQuery
  }
}

class PelagoRelMetadataQuery protected () extends RelMetadataQuery {
  private var deviceTypeHandler = initialHandler(classOf[DeviceType.Handler])
  private var packingHandler = initialHandler(classOf[Packing.Handler])
  //    selfCostHandler    = initialHandler(SelfCost  .Handler.class);
  private var hetDistrHandler = initialHandler(classOf[HetDistribution.Handler])
  private var homDistrHandler = initialHandler(classOf[HomDistribution.Handler])
  private var computeTypeHandler = initialHandler(
    classOf[ComputeDevice.Handler]
  )

  def deviceType(rel: RelNode): RelDeviceType = {
    while (true)
      try return deviceTypeHandler.deviceType(rel, this)
      catch {
        case e: JaninoRelMetadataProvider.NoHandler =>
          deviceTypeHandler = revise(e.relClass, DeviceType.DEF)
      }
    null
  }

  def hetDistribution(rel: RelNode): RelHetDistribution = {
    while (true)
      try return hetDistrHandler.hetDistribution(rel, this)
      catch {
        case e: JaninoRelMetadataProvider.NoHandler =>
          hetDistrHandler = revise(e.relClass, HetDistribution.DEF)
      }
    null
  }

  def homDistribution(rel: RelNode): RelHomDistribution = {
    while (true)
      try return homDistrHandler.homDistribution(rel, this)
      catch {
        case e: JaninoRelMetadataProvider.NoHandler =>
          homDistrHandler = revise(e.relClass, HomDistribution.DEF)
      }
    null
  }

  override def distribution(rel: RelNode): RelDistribution = {
    assert(false)
    null
  }

  def computeType(rel: RelNode): RelComputeDevice = {
    while (true)
      try return computeTypeHandler.computeType(rel, this)
      catch {
        case e: JaninoRelMetadataProvider.NoHandler =>
          computeTypeHandler = revise(e.relClass, ComputeDevice.DEF)
      }
    null
  }

  def packing(rel: RelNode): RelPacking = {
    while (true)
      try return packingHandler.packing(rel, this)
      catch {
        case e: JaninoRelMetadataProvider.NoHandler =>
          packingHandler = revise(e.relClass, Packing.DEF)
      }
    null
  }
}
